fastplotlib/layouts/_record_mixin.py

```diff
@@ -0,0 +1,238 @@
+from pathlib import Path
+from multiprocessing import Queue, Process
+from time import time
+
+try:
+    import av
+except ImportError:
+    HAS_AV = False
+else:
+    HAS_AV = True
+
+
+class VideoWriterAV(Process):
+    """Video writer, uses PyAV in an external process to write frames to disk"""
+    def __init__(
+            self,
+            path: Union[Path, str],
+            queue: Queue,
+            fps: int,
+            width: int,
+            height: int,
+            codec: str,
+            pixel_format: str,
+            options: dict = None
+    ):
+        super().__init__()
+        self.queue = queue
+
+        self.container = av.open(path, mode="w")
+
+        self.stream = self.container.add_stream(codec, rate=fps, options=options)
+
+        # in case libx264, trim last rows and/or column
+        # because libx264 doesn't like non-even number width or height
+        if width % 2 != 0:
+            width -= 1
+        if height % 2 != 0:
+            height -= 1
+
+        self.stream.width = width
+        self.stream.height = height
+
+        self.stream.pix_fmt = pixel_format
+
+    def run(self):
+        while True:
+            if self.queue.empty():  # no frame to write
+                continue
+
+            frame = self.queue.get()
+
+            # recording has ended
+            if frame is None:
+                self.container.close()
+                break
+
+            frame = av.VideoFrame.from_ndarray(
+                frame[:self.stream.height, :self.stream.width],  # trim if necessary because of x264
+                format="rgb24"
+            )
+
+            for packet in self.stream.encode(frame):
+                self.container.mux(packet)
+
+        # I don't exactly know what this does, copied from pyav example
+        for packet in self.stream.encode():
+            self.container.mux(packet)
+
+        # close file
+        self.container.close()
+
+        # close process, release resources
+        self.close()
+
+
+# adds recording functionality to GridPlot and Plot
+class RecordMixin:
+    def __init__(self):
+        self._video_writer: VideoWriterAV = None
+        self._video_writer_queue = Queue()
+        self._record_fps = 25
+        self._record_timer = 0
+        self._record_start_time = 0
+
+    def _record(self):
+        """
+        Sends frame to VideoWriter through video writer queue
+        """
+        # current time
+        t = time()
+
+        # put frame in queue only if enough time as passed according to the desired framerate
+        # otherwise it tries to record EVERY frame on every rendering cycle, which just blocks the rendering
+        if t - self._record_timer < (1 / self._record_fps):
+            return
+
+        # reset timer
+        self._record_timer = t
+
+        if self._video_writer is not None:
+            ss = self.canvas.snapshot()
+            # exclude alpha channel
+            self._video_writer_queue.put(ss.data[..., :-1])
+
```

Comment:
what indexing does this do on the snapshot? 

`ss.data[..., :-1]`

Reply:
`...` is to skip dimensions, so this way we only index the last dimension (RGBA) and we then only take the RGB, discard alpha channel.

Equivalent to `[:, :, - 1]`